Convert color types to strongly typed CSS/Style values

weinig opened the pull request:
#### 0f12ab8938f6d250560003ba3e43215aae7c0934
<pre>
Convert color types to strongly typed CSS/Style values
<a href="https://bugs.webkit.org/show_bug.cgi?id=283828">https://bugs.webkit.org/show_bug.cgi?id=283828</a>

Reviewed by Darin Adler.

Integrates color types into the strongly typed CSS/Style value architecture.

- Renames all CSSUnresolved*Foo color types to CSS::Foo and moves them to css/values/color/.
- Renames all Style*Foo color types to Style::Foo and moves them to style/values/color/.
- Swaps which types use UniqueRefs in CSS::Color to match Style::Color. This is needed
  because now CSS color values are stored in a new CSSColorValue type instead of being
  allocated and and referenced from a CSSPrimitiveValue. The CSSValuePool now vends/caches
  CSSColorValues for colors created through it.
- Implements standard interfaces for strongly type CSS/Style values for color types
  (e.g. Serialization&lt;&gt;, ComputedStyleDependenciesCollector&lt;&gt;, ToStyle&lt;&gt;/ToCSS&lt;&gt;, etc.).
- Moves color keyword lookup to CSSKeywordColor from StyleColor.
- Adopts CSS::Color in CSS::GradientColorStop to prove things out. To better model the
  semantics of color stops, the color value is now generic, allowing Angular/Linear
  stops to use Markable&lt;CSS::Color&gt; whereas Deprecated stops use CSS::Color directly,
  since they are required at parse time.
- To avoid name collisions, the typedom CSS color types where updated to use the prefix
  CSSOM rather than CSS, matching precedent set in CSSOMVariableReferenceValue.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/css/CSSColorValue.cpp: Added.
* Source/WebCore/css/CSSColorValue.h: Added.
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/CSSFontPaletteValuesOverrideColorsValue.h:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSShadowValue.cpp:
* Source/WebCore/css/CSSShadowValue.h:
* Source/WebCore/css/CSSUnits.cpp:
* Source/WebCore/css/CSSUnits.h:
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/CSSValuePool.cpp:
* Source/WebCore/css/CSSValuePool.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/ComputedStyleExtractor.h:
* Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.cpp:
* Source/WebCore/css/DeprecatedCSSOMRGBColor.h:
* Source/WebCore/css/SVGCSSComputedStyleDeclaration.cpp:
* Source/WebCore/css/StyleProperties.cpp:
* Source/WebCore/css/calc/CSSCalcTree+NumericIdentity.h:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree.cpp:
* Source/WebCore/css/calc/CSSCalcType.cpp:
* Source/WebCore/css/parser/CSSParser.cpp:
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Background.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.cpp:
* Source/WebCore/css/process-css-properties.py:
* Source/WebCore/css/query/MediaQueryEvaluator.cpp:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
* Source/WebCore/css/typedom/color/CSSHSL.h:
* Source/WebCore/css/typedom/color/CSSHSL.idl:
* Source/WebCore/css/typedom/color/CSSHWB.h:
* Source/WebCore/css/typedom/color/CSSHWB.idl:
* Source/WebCore/css/typedom/color/CSSLCH.h:
* Source/WebCore/css/typedom/color/CSSLCH.idl:
* Source/WebCore/css/typedom/color/CSSLab.h:
* Source/WebCore/css/typedom/color/CSSLab.idl:
* Source/WebCore/css/typedom/color/CSSOKLCH.h:
* Source/WebCore/css/typedom/color/CSSOKLCH.idl:
* Source/WebCore/css/typedom/color/CSSOKLab.h:
* Source/WebCore/css/typedom/color/CSSOKLab.idl:
* Source/WebCore/css/typedom/color/CSSOMColor.cpp: Renamed from Source/WebCore/css/typedom/color/CSSColor.cpp.
* Source/WebCore/css/typedom/color/CSSOMColor.h: Renamed from Source/WebCore/css/typedom/color/CSSColor.h.
* Source/WebCore/css/typedom/color/CSSOMColor.idl: Renamed from Source/WebCore/css/typedom/color/CSSColor.idl.
* Source/WebCore/css/typedom/color/CSSOMColorValue.cpp: Renamed from Source/WebCore/css/typedom/color/CSSColorValue.cpp.
* Source/WebCore/css/typedom/color/CSSOMColorValue.h: Renamed from Source/WebCore/css/typedom/color/CSSColorValue.h.
* Source/WebCore/css/typedom/color/CSSOMColorValue.idl: Renamed from Source/WebCore/css/typedom/color/CSSColorValue.idl.
* Source/WebCore/css/typedom/color/CSSRGB.h:
* Source/WebCore/css/typedom/color/CSSRGB.idl:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/css/values/color/CSSAbsoluteColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedAbsoluteColor.h.
* Source/WebCore/css/values/color/CSSAbsoluteColorResolver.h: Renamed from Source/WebCore/css/color/CSSAbsoluteColorResolver.h.
* Source/WebCore/css/values/color/CSSAbsoluteColorSerialization.h: Renamed from Source/WebCore/css/color/CSSAbsoluteColorSerialization.h.
* Source/WebCore/css/values/color/CSSColor.cpp: Renamed from Source/WebCore/css/color/CSSUnresolvedColor.cpp.
* Source/WebCore/css/values/color/CSSColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedColor.h.
* Source/WebCore/css/values/color/CSSColorConversion+Normalize.h: Renamed from Source/WebCore/css/color/CSSColorConversion+Normalize.h.
* Source/WebCore/css/values/color/CSSColorConversion+ToColor.h: Renamed from Source/WebCore/css/color/CSSColorConversion+ToColor.h.
* Source/WebCore/css/values/color/CSSColorConversion+ToTypedColor.h: Renamed from Source/WebCore/css/color/CSSColorConversion+ToTypedColor.h.
* Source/WebCore/css/values/color/CSSColorDescriptors.h: Renamed from Source/WebCore/css/color/CSSColorDescriptors.h.
* Source/WebCore/css/values/color/CSSColorLayers.cpp: Renamed from Source/WebCore/css/color/CSSUnresolvedColorLayers.cpp.
* Source/WebCore/css/values/color/CSSColorLayers.h: Renamed from Source/WebCore/css/color/CSSUnresolvedColorLayers.h.
* Source/WebCore/css/values/color/CSSColorLayersResolver.cpp: Renamed from Source/WebCore/css/color/CSSColorLayersResolver.cpp.
* Source/WebCore/css/values/color/CSSColorLayersResolver.h: Copied from Source/WebCore/css/color/CSSColorLayersResolver.h.
* Source/WebCore/css/values/color/CSSColorLayersSerialization.h: Renamed from Source/WebCore/css/color/CSSColorLayersSerialization.h.
* Source/WebCore/css/values/color/CSSColorMix.cpp: Renamed from Source/WebCore/css/color/CSSUnresolvedColorMix.cpp.
* Source/WebCore/css/values/color/CSSColorMix.h: Renamed from Source/WebCore/css/color/CSSUnresolvedColorMix.h.
* Source/WebCore/css/values/color/CSSColorMixResolver.cpp: Renamed from Source/WebCore/css/color/CSSColorMixResolver.cpp.
* Source/WebCore/css/values/color/CSSColorMixResolver.h: Renamed from Source/WebCore/css/color/CSSColorMixResolver.h.
* Source/WebCore/css/values/color/CSSColorMixSerialization.cpp: Renamed from Source/WebCore/css/color/CSSColorMixSerialization.cpp.
* Source/WebCore/css/values/color/CSSColorMixSerialization.h: Renamed from Source/WebCore/css/color/CSSColorMixSerialization.h.
* Source/WebCore/css/values/color/CSSColorType.h: Copied from Source/WebCore/css/color/CSSUnresolvedRelativeColor.cpp.
* Source/WebCore/css/values/color/CSSContrastColor.cpp: Copied from Source/WebCore/css/color/StyleContrastColor.cpp.
* Source/WebCore/css/values/color/CSSContrastColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedContrastColor.h.
* Source/WebCore/css/values/color/CSSContrastColorResolver.cpp: Renamed from Source/WebCore/css/color/CSSContrastColorResolver.cpp.
* Source/WebCore/css/values/color/CSSContrastColorResolver.h: Renamed from Source/WebCore/css/color/CSSContrastColorResolver.h.
* Source/WebCore/css/values/color/CSSContrastColorSerialization.h: Renamed from Source/WebCore/css/color/CSSContrastColorSerialization.h.
* Source/WebCore/css/values/color/CSSHexColor.cpp: Renamed from Source/WebCore/css/color/CSSUnresolvedColorHex.cpp.
* Source/WebCore/css/values/color/CSSHexColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedColorHex.h.
* Source/WebCore/css/values/color/CSSKeywordColor.cpp: Renamed from Source/WebCore/css/color/CSSUnresolvedColorKeyword.cpp.
* Source/WebCore/css/values/color/CSSKeywordColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedColorKeyword.h.
* Source/WebCore/css/values/color/CSSLightDarkColor.cpp: Renamed from Source/WebCore/css/color/CSSUnresolvedLightDark.cpp.
* Source/WebCore/css/values/color/CSSLightDarkColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedLightDark.h.
* Source/WebCore/css/values/color/CSSPlatformColorResolutionState.cpp: Added.
* Source/WebCore/css/values/color/CSSPlatformColorResolutionState.h: Renamed from Source/WebCore/css/color/CSSUnresolvedColorResolutionState.h.
* Source/WebCore/css/values/color/CSSRelativeColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedRelativeColor.h.
* Source/WebCore/css/values/color/CSSRelativeColorResolver.h: Renamed from Source/WebCore/css/color/CSSRelativeColorResolver.h.
* Source/WebCore/css/values/color/CSSRelativeColorSerialization.h: Renamed from Source/WebCore/css/color/CSSRelativeColorSerialization.h.
* Source/WebCore/css/values/color/CSSResolvedColor.cpp: Renamed from Source/WebCore/css/color/CSSUnresolvedAbsoluteResolvedColor.cpp.
* Source/WebCore/css/values/color/CSSResolvedColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedAbsoluteResolvedColor.h.
* Source/WebCore/css/values/images/CSSGradient.cpp:
* Source/WebCore/css/values/images/CSSGradient.h:
* Source/WebCore/css/values/primitives/CSSNone.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+ComputedStyleDependencies.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes.cpp:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/editing/EditingStyle.cpp:
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
* Source/WebCore/html/ColorInputType.cpp:
* Source/WebCore/html/HTMLInputElement.cpp:
* Source/WebCore/html/canvas/CanvasStyle.cpp:
* Source/WebCore/rendering/RenderTreeAsText.cpp:
* Source/WebCore/rendering/style/BorderValue.cpp:
* Source/WebCore/rendering/style/BorderValue.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/SVGRenderStyle.cpp:
* Source/WebCore/rendering/style/SVGRenderStyle.h:
* Source/WebCore/rendering/style/SVGRenderStyleDefs.h:
* Source/WebCore/rendering/style/ScrollbarColor.h:
* Source/WebCore/rendering/style/ShadowData.h:
* Source/WebCore/rendering/style/StyleBackgroundData.h:
* Source/WebCore/rendering/style/StyleMultiColData.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleVisitedLinkColorData.h:
* Source/WebCore/rendering/svg/SVGPaintServerHandling.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResource.cpp:
* Source/WebCore/style/ColorFromPrimitiveValue.cpp: Removed.
* Source/WebCore/style/ColorFromPrimitiveValue.h: Removed.
* Source/WebCore/style/FilterOperationsBuilder.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleForVisitedLink.h: Added.
* Source/WebCore/style/StyleResolver.cpp:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/color/StyleAbsoluteColor.h: Copied from Source/WebCore/css/color/StyleAbsoluteColor.h.
* Source/WebCore/style/values/color/StyleColor.cpp: Renamed from Source/WebCore/css/color/StyleColor.cpp..
* Source/WebCore/style/values/color/StyleColor.h: Renamed from Source/WebCore/css/color/StyleColor.h..
* Source/WebCore/style/values/color/StyleColorLayers.cpp: Renamed from Source/WebCore/css/color/StyleColorLayers.cpp.
* Source/WebCore/style/values/color/StyleColorLayers.h: Renamed from Source/WebCore/css/color/StyleColorLayers.h.
* Source/WebCore/style/values/color/StyleColorMix.cpp: Renamed from Source/WebCore/css/color/StyleColorMix.cpp.
* Source/WebCore/style/values/color/StyleColorMix.h: Renamed from Source/WebCore/css/color/StyleColorMix.h.
* Source/WebCore/style/values/color/StyleColorOptions.cpp: Added.
* Source/WebCore/style/values/color/StyleColorOptions.h: Added.
* Source/WebCore/style/values/color/StyleColorResolutionState.h: Renamed from Source/WebCore/css/color/CSSUnresolvedStyleColorResolutionState.h.
* Source/WebCore/style/values/color/StyleContrastColor.cpp: Renamed from Source/WebCore/css/color/StyleContrastColor.cpp.
* Source/WebCore/style/values/color/StyleContrastColor.h: Renamed from Source/WebCore/css/color/StyleContrastColor.h.
* Source/WebCore/style/values/color/StyleCurrentColor.cpp: Copied from Source/WebCore/css/color/StyleCurrentColor.cpp.
* Source/WebCore/style/values/color/StyleCurrentColor.h: Renamed from Source/WebCore/css/color/StyleCurrentColor.h.
* Source/WebCore/style/values/color/StyleHexColor.cpp: Added.
* Source/WebCore/style/values/color/StyleHexColor.h: Added.
* Source/WebCore/style/values/color/StyleKeywordColor.cpp: Added.
* Source/WebCore/style/values/color/StyleKeywordColor.h: Added.
* Source/WebCore/style/values/color/StyleLightDarkColor.cpp: Added.
* Source/WebCore/style/values/color/StyleLightDarkColor.h: Added.
* Source/WebCore/style/values/color/StyleRelativeColor.h: Renamed from Source/WebCore/css/color/StyleRelativeColor.h.
* Source/WebCore/style/values/color/StyleResolvedColor.cpp: Renamed from Source/WebCore/css/color/StyleAbsoluteColor.cpp.
* Source/WebCore/style/values/color/StyleResolvedColor.h: Renamed from Source/WebCore/css/color/StyleAbsoluteColor.h.
* Source/WebCore/style/values/images/StyleGradient.cpp:
* Source/WebCore/style/values/images/StyleGradient.h:
* Source/WebCore/testing/Internals.cpp:

Canonical link: <a href="https://commits.webkit.org/287234@main">https://commits.webkit.org/287234@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/aa5248bbf6022fa9fe74dcd4f0f3a3903ed5edec

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78901 "73 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57945 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32283 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83554 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30150 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81034 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67083 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6226 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/61784 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/19709 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81968 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51822 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/71792 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42088 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/49172 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/25995 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28498 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/70286 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26404 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84930 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6257 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/4320 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/70020 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6420 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67824 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69271 "Found 4 new API test failures: /TestWebKit:WebKit2.ProvisionalURLAfterWillSendRequestCallback, /TestWebKit:WebKit.GeolocationBasicWithHighAccuracy, /TestWebKit:WebKit.FocusedFrameAfterCrash, /TestWebKit:WebKit.Find (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17244 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13315 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/12047 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6205 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/12073 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/125/builds/6177 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9619 "Built successfully") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/124/builds/7968 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WebCore/css/CSSColorValue.h:
```diff
@@ -16,7 +16,6 @@
  * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
  * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
  * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
- * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
```
Why delete this line? Presumably just an accident?

weinig replied:
Yeah, accident. Will fix.

weinig replied:
Wow! 324 files in WebCore are missing that line!

darinadler commented on Source/WebCore/css/CSSColorValue.h:
```diff
@@ -25,31 +24,38 @@
 
-#include <wtf/Forward.h>
+#include "CSSColor.h"
+#include "CSSValue.h"
```
Why can't this be forward declarations instead?

weinig replied:
I believe you need to include the full type for both of these:

CSSValue is needed because CSSColorValue inherits from it.
CSSColor is needed because CSSColorValue has a CSS::Color as a member (non-pointer member).

darinadler commented on Source/WebCore/css/CSSColorValue.cpp:
```diff
@@ -0,0 +1,83 @@
+        return color->color().absoluteColor();
+
+    auto valueID = value.valueID();
+    if (CSS::isAbsoluteColorKeyword(valueID))
```
```suggestion
    if (auto valueID = value.valueID(); CSS::isAbsoluteColorKeyword(valueID))
```

darinadler commented on Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
```diff
@@ -266,14 +267,19 @@ static RefPtr<CSSValue> consumeDeprecatedGradient(CSSParserTokenRange& range, co
 
 enum class SupportsColorHints : bool { No, Yes };
 
+static Markable<CSS::Color> consumeStopColor(CSSParserTokenRange& range, const CSSParserContext& context)
```
I thought we usually use `std::optional` right up until we store in a `Markable` data member?

darinadler commented on Source/WebCore/css/values/color/CSSKeywordColor.cpp:
```diff
@@ -0,0 +1,144 @@
+{
+    // https://drafts.csswg.org/css-color-4/#named-colors
+    // "16 of CSS’s named colors come from the VGA palette originally, and were then adopted into HTML"
+    return (id >= CSSValueAqua && id <= CSSValueGrey);
```
No need for the parentheses

darinadler commented on Source/WebCore/style/values/color/StyleColorResolutionState.h:
```diff
@@ -27,37 +27,40 @@
+
+enum class ForVisitedLink : bool;
+
+// Used to resolves a CSS::Color to a Style::Color during style building.
```
```suggestion
// Used to resolve a CSS::Color to a Style::Color during style building.
```

darinadler commented on Source/WebCore/style/values/color/StyleKeywordColor.h:
```diff
@@ -0,0 +1,41 @@
+
+#pragma once
+
+#include "CSSKeywordColor.h"
```
Can't this be a forward declaration instead?

darinadler commented on Source/WebCore/style/values/color/StyleLightDarkColor.h:
```diff
@@ -0,0 +1,41 @@
+
+#pragma once
+
+#include "CSSLightDarkColor.h"
```
Can't this be a forward declaration instead?

weinig committed:
Convert color types to strongly typed CSS/Style values
https://bugs.webkit.org/show_bug.cgi?id=283828

Reviewed by Darin Adler.

Integrates color types into the strongly typed CSS/Style value architecture.

- Renames all CSSUnresolved*Foo color types to CSS::Foo and moves them to css/values/color/.
- Renames all Style*Foo color types to Style::Foo and moves them to style/values/color/.
- Swaps which types use UniqueRefs in CSS::Color to match Style::Color. This is needed
  because now CSS color values are stored in a new CSSColorValue type instead of being
  allocated and and referenced from a CSSPrimitiveValue. The CSSValuePool now vends/caches
  CSSColorValues for colors created through it.
- Implements standard interfaces for strongly type CSS/Style values for color types
  (e.g. Serialization<>, ComputedStyleDependenciesCollector<>, ToStyle<>/ToCSS<>, etc.).
- Moves color keyword lookup to CSSKeywordColor from StyleColor.
- Adopts CSS::Color in CSS::GradientColorStop to prove things out. To better model the
  semantics of color stops, the color value is now generic, allowing Angular/Linear
  stops to use Markable<CSS::Color> whereas Deprecated stops use CSS::Color directly,
  since they are required at parse time.
- To avoid name collisions, the typedom CSS color types where updated to use the prefix
  CSSOM rather than CSS, matching precedent set in CSSOMVariableReferenceValue.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/SaferCPPExpectations/NoUncountedMemberCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/css/CSSColorValue.cpp: Added.
* Source/WebCore/css/CSSColorValue.h: Added.
* Source/WebCore/css/CSSCustomPropertyValue.cpp:
* Source/WebCore/css/CSSCustomPropertyValue.h:
* Source/WebCore/css/CSSFontPaletteValuesOverrideColorsValue.h:
* Source/WebCore/css/CSSGradientValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.cpp:
* Source/WebCore/css/CSSPrimitiveValue.h:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSShadowValue.cpp:
* Source/WebCore/css/CSSShadowValue.h:
* Source/WebCore/css/CSSUnits.cpp:
* Source/WebCore/css/CSSUnits.h:
* Source/WebCore/css/CSSValue.cpp:
* Source/WebCore/css/CSSValue.h:
* Source/WebCore/css/CSSValuePool.cpp:
* Source/WebCore/css/CSSValuePool.h:
* Source/WebCore/css/ComputedStyleExtractor.cpp:
* Source/WebCore/css/ComputedStyleExtractor.h:
* Source/WebCore/css/DeprecatedCSSOMPrimitiveValue.cpp:
* Source/WebCore/css/DeprecatedCSSOMRGBColor.h:
* Source/WebCore/css/SVGCSSComputedStyleDeclaration.cpp:
* Source/WebCore/css/StyleProperties.cpp:
* Source/WebCore/css/calc/CSSCalcTree+NumericIdentity.h:
* Source/WebCore/css/calc/CSSCalcTree+Serialization.cpp:
* Source/WebCore/css/calc/CSSCalcTree+Simplification.cpp:
* Source/WebCore/css/calc/CSSCalcTree.cpp:
* Source/WebCore/css/calc/CSSCalcType.cpp:
* Source/WebCore/css/parser/CSSParser.cpp:
* Source/WebCore/css/parser/CSSParser.h:
* Source/WebCore/css/parser/CSSParserContext.cpp:
* Source/WebCore/css/parser/CSSParserContext.h:
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
* Source/WebCore/css/parser/CSSParserImpl.cpp:
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Background.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Color.h:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Filter.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Font.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+Image.cpp:
* Source/WebCore/css/parser/CSSPropertyParserConsumer+TextDecoration.cpp:
* Source/WebCore/css/process-css-properties.py:
* Source/WebCore/css/query/MediaQueryEvaluator.cpp:
* Source/WebCore/css/typedom/CSSStyleValueFactory.cpp:
* Source/WebCore/css/typedom/color/CSSHSL.h:
* Source/WebCore/css/typedom/color/CSSHSL.idl:
* Source/WebCore/css/typedom/color/CSSHWB.h:
* Source/WebCore/css/typedom/color/CSSHWB.idl:
* Source/WebCore/css/typedom/color/CSSLCH.h:
* Source/WebCore/css/typedom/color/CSSLCH.idl:
* Source/WebCore/css/typedom/color/CSSLab.h:
* Source/WebCore/css/typedom/color/CSSLab.idl:
* Source/WebCore/css/typedom/color/CSSOKLCH.h:
* Source/WebCore/css/typedom/color/CSSOKLCH.idl:
* Source/WebCore/css/typedom/color/CSSOKLab.h:
* Source/WebCore/css/typedom/color/CSSOKLab.idl:
* Source/WebCore/css/typedom/color/CSSOMColor.cpp: Renamed from Source/WebCore/css/typedom/color/CSSColor.cpp.
* Source/WebCore/css/typedom/color/CSSOMColor.h: Renamed from Source/WebCore/css/typedom/color/CSSColor.h.
* Source/WebCore/css/typedom/color/CSSOMColor.idl: Renamed from Source/WebCore/css/typedom/color/CSSColor.idl.
* Source/WebCore/css/typedom/color/CSSOMColorValue.cpp: Renamed from Source/WebCore/css/typedom/color/CSSColorValue.cpp.
* Source/WebCore/css/typedom/color/CSSOMColorValue.h: Renamed from Source/WebCore/css/typedom/color/CSSColorValue.h.
* Source/WebCore/css/typedom/color/CSSOMColorValue.idl: Renamed from Source/WebCore/css/typedom/color/CSSColorValue.idl.
* Source/WebCore/css/typedom/color/CSSRGB.h:
* Source/WebCore/css/typedom/color/CSSRGB.idl:
* Source/WebCore/css/values/CSSValueTypes.h:
* Source/WebCore/css/values/color/CSSAbsoluteColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedAbsoluteColor.h.
* Source/WebCore/css/values/color/CSSAbsoluteColorResolver.h: Renamed from Source/WebCore/css/color/CSSAbsoluteColorResolver.h.
* Source/WebCore/css/values/color/CSSAbsoluteColorSerialization.h: Renamed from Source/WebCore/css/color/CSSAbsoluteColorSerialization.h.
* Source/WebCore/css/values/color/CSSColor.cpp: Renamed from Source/WebCore/css/color/CSSUnresolvedColor.cpp.
* Source/WebCore/css/values/color/CSSColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedColor.h.
* Source/WebCore/css/values/color/CSSColorConversion+Normalize.h: Renamed from Source/WebCore/css/color/CSSColorConversion+Normalize.h.
* Source/WebCore/css/values/color/CSSColorConversion+ToColor.h: Renamed from Source/WebCore/css/color/CSSColorConversion+ToColor.h.
* Source/WebCore/css/values/color/CSSColorConversion+ToTypedColor.h: Renamed from Source/WebCore/css/color/CSSColorConversion+ToTypedColor.h.
* Source/WebCore/css/values/color/CSSColorDescriptors.h: Renamed from Source/WebCore/css/color/CSSColorDescriptors.h.
* Source/WebCore/css/values/color/CSSColorLayers.cpp: Renamed from Source/WebCore/css/color/CSSUnresolvedColorLayers.cpp.
* Source/WebCore/css/values/color/CSSColorLayers.h: Renamed from Source/WebCore/css/color/CSSUnresolvedColorLayers.h.
* Source/WebCore/css/values/color/CSSColorLayersResolver.cpp: Renamed from Source/WebCore/css/color/CSSColorLayersResolver.cpp.
* Source/WebCore/css/values/color/CSSColorLayersResolver.h: Copied from Source/WebCore/css/color/CSSColorLayersResolver.h.
* Source/WebCore/css/values/color/CSSColorLayersSerialization.h: Renamed from Source/WebCore/css/color/CSSColorLayersSerialization.h.
* Source/WebCore/css/values/color/CSSColorMix.cpp: Renamed from Source/WebCore/css/color/CSSUnresolvedColorMix.cpp.
* Source/WebCore/css/values/color/CSSColorMix.h: Renamed from Source/WebCore/css/color/CSSUnresolvedColorMix.h.
* Source/WebCore/css/values/color/CSSColorMixResolver.cpp: Renamed from Source/WebCore/css/color/CSSColorMixResolver.cpp.
* Source/WebCore/css/values/color/CSSColorMixResolver.h: Renamed from Source/WebCore/css/color/CSSColorMixResolver.h.
* Source/WebCore/css/values/color/CSSColorMixSerialization.cpp: Renamed from Source/WebCore/css/color/CSSColorMixSerialization.cpp.
* Source/WebCore/css/values/color/CSSColorMixSerialization.h: Renamed from Source/WebCore/css/color/CSSColorMixSerialization.h.
* Source/WebCore/css/values/color/CSSColorType.h: Copied from Source/WebCore/css/color/CSSUnresolvedRelativeColor.cpp.
* Source/WebCore/css/values/color/CSSContrastColor.cpp: Copied from Source/WebCore/css/color/StyleContrastColor.cpp.
* Source/WebCore/css/values/color/CSSContrastColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedContrastColor.h.
* Source/WebCore/css/values/color/CSSContrastColorResolver.cpp: Renamed from Source/WebCore/css/color/CSSContrastColorResolver.cpp.
* Source/WebCore/css/values/color/CSSContrastColorResolver.h: Renamed from Source/WebCore/css/color/CSSContrastColorResolver.h.
* Source/WebCore/css/values/color/CSSContrastColorSerialization.h: Renamed from Source/WebCore/css/color/CSSContrastColorSerialization.h.
* Source/WebCore/css/values/color/CSSHexColor.cpp: Renamed from Source/WebCore/css/color/CSSUnresolvedColorHex.cpp.
* Source/WebCore/css/values/color/CSSHexColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedColorHex.h.
* Source/WebCore/css/values/color/CSSKeywordColor.cpp: Renamed from Source/WebCore/css/color/CSSUnresolvedColorKeyword.cpp.
* Source/WebCore/css/values/color/CSSKeywordColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedColorKeyword.h.
* Source/WebCore/css/values/color/CSSLightDarkColor.cpp: Renamed from Source/WebCore/css/color/CSSUnresolvedLightDark.cpp.
* Source/WebCore/css/values/color/CSSLightDarkColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedLightDark.h.
* Source/WebCore/css/values/color/CSSPlatformColorResolutionState.cpp: Added.
* Source/WebCore/css/values/color/CSSPlatformColorResolutionState.h: Renamed from Source/WebCore/css/color/CSSUnresolvedColorResolutionState.h.
* Source/WebCore/css/values/color/CSSRelativeColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedRelativeColor.h.
* Source/WebCore/css/values/color/CSSRelativeColorResolver.h: Renamed from Source/WebCore/css/color/CSSRelativeColorResolver.h.
* Source/WebCore/css/values/color/CSSRelativeColorSerialization.h: Renamed from Source/WebCore/css/color/CSSRelativeColorSerialization.h.
* Source/WebCore/css/values/color/CSSResolvedColor.cpp: Renamed from Source/WebCore/css/color/CSSUnresolvedAbsoluteResolvedColor.cpp.
* Source/WebCore/css/values/color/CSSResolvedColor.h: Renamed from Source/WebCore/css/color/CSSUnresolvedAbsoluteResolvedColor.h.
* Source/WebCore/css/values/images/CSSGradient.cpp:
* Source/WebCore/css/values/images/CSSGradient.h:
* Source/WebCore/css/values/primitives/CSSNone.h:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes+ComputedStyleDependencies.cpp:
* Source/WebCore/css/values/primitives/CSSPrimitiveNumericTypes.cpp:
* Source/WebCore/dom/Document.cpp:
* Source/WebCore/editing/EditingStyle.cpp:
* Source/WebCore/editing/cocoa/HTMLConverter.mm:
* Source/WebCore/html/ColorInputType.cpp:
* Source/WebCore/html/HTMLInputElement.cpp:
* Source/WebCore/html/canvas/CanvasStyle.cpp:
* Source/WebCore/rendering/RenderTreeAsText.cpp:
* Source/WebCore/rendering/style/BorderValue.cpp:
* Source/WebCore/rendering/style/BorderValue.h:
* Source/WebCore/rendering/style/RenderStyle.cpp:
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleInlines.h:
* Source/WebCore/rendering/style/RenderStyleSetters.h:
* Source/WebCore/rendering/style/SVGRenderStyle.cpp:
* Source/WebCore/rendering/style/SVGRenderStyle.h:
* Source/WebCore/rendering/style/SVGRenderStyleDefs.h:
* Source/WebCore/rendering/style/ScrollbarColor.h:
* Source/WebCore/rendering/style/ShadowData.h:
* Source/WebCore/rendering/style/StyleBackgroundData.h:
* Source/WebCore/rendering/style/StyleMultiColData.h:
* Source/WebCore/rendering/style/StyleRareInheritedData.cpp:
* Source/WebCore/rendering/style/StyleRareInheritedData.h:
* Source/WebCore/rendering/style/StyleRareNonInheritedData.h:
* Source/WebCore/rendering/style/StyleVisitedLinkColorData.h:
* Source/WebCore/rendering/svg/SVGPaintServerHandling.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGResource.cpp:
* Source/WebCore/style/ColorFromPrimitiveValue.cpp: Removed.
* Source/WebCore/style/ColorFromPrimitiveValue.h: Removed.
* Source/WebCore/style/FilterOperationsBuilder.cpp:
* Source/WebCore/style/StyleBuilderConverter.h:
* Source/WebCore/style/StyleBuilderCustom.h:
* Source/WebCore/style/StyleBuilderState.cpp:
* Source/WebCore/style/StyleBuilderState.h:
* Source/WebCore/style/StyleForVisitedLink.h: Added.
* Source/WebCore/style/StyleResolver.cpp:
* Source/WebCore/style/values/StyleValueTypes.h:
* Source/WebCore/style/values/color/StyleAbsoluteColor.h: Copied from Source/WebCore/css/color/StyleAbsoluteColor.h.
* Source/WebCore/style/values/color/StyleColor.cpp: Renamed from Source/WebCore/css/color/StyleColor.cpp..
* Source/WebCore/style/values/color/StyleColor.h: Renamed from Source/WebCore/css/color/StyleColor.h..
* Source/WebCore/style/values/color/StyleColorLayers.cpp: Renamed from Source/WebCore/css/color/StyleColorLayers.cpp.
* Source/WebCore/style/values/color/StyleColorLayers.h: Renamed from Source/WebCore/css/color/StyleColorLayers.h.
* Source/WebCore/style/values/color/StyleColorMix.cpp: Renamed from Source/WebCore/css/color/StyleColorMix.cpp.
* Source/WebCore/style/values/color/StyleColorMix.h: Renamed from Source/WebCore/css/color/StyleColorMix.h.
* Source/WebCore/style/values/color/StyleColorOptions.cpp: Added.
* Source/WebCore/style/values/color/StyleColorOptions.h: Added.
* Source/WebCore/style/values/color/StyleColorResolutionState.h: Renamed from Source/WebCore/css/color/CSSUnresolvedStyleColorResolutionState.h.
* Source/WebCore/style/values/color/StyleContrastColor.cpp: Renamed from Source/WebCore/css/color/StyleContrastColor.cpp.
* Source/WebCore/style/values/color/StyleContrastColor.h: Renamed from Source/WebCore/css/color/StyleContrastColor.h.
* Source/WebCore/style/values/color/StyleCurrentColor.cpp: Copied from Source/WebCore/css/color/StyleCurrentColor.cpp.
* Source/WebCore/style/values/color/StyleCurrentColor.h: Renamed from Source/WebCore/css/color/StyleCurrentColor.h.
* Source/WebCore/style/values/color/StyleHexColor.cpp: Added.
* Source/WebCore/style/values/color/StyleHexColor.h: Added.
* Source/WebCore/style/values/color/StyleKeywordColor.cpp: Added.
* Source/WebCore/style/values/color/StyleKeywordColor.h: Added.
* Source/WebCore/style/values/color/StyleLightDarkColor.cpp: Added.
* Source/WebCore/style/values/color/StyleLightDarkColor.h: Added.
* Source/WebCore/style/values/color/StyleRelativeColor.h: Renamed from Source/WebCore/css/color/StyleRelativeColor.h.
* Source/WebCore/style/values/color/StyleResolvedColor.cpp: Renamed from Source/WebCore/css/color/StyleAbsoluteColor.cpp.
* Source/WebCore/style/values/color/StyleResolvedColor.h: Renamed from Source/WebCore/css/color/StyleAbsoluteColor.h.
* Source/WebCore/style/values/images/StyleGradient.cpp:
* Source/WebCore/style/values/images/StyleGradient.h:
* Source/WebCore/testing/Internals.cpp:

Canonical link: https://commits.webkit.org/287234@main
